ogs6py/log_parser/common_ogs_analyses.py

```diff
@@ -0,0 +1,176 @@
+
+    dfe_ts = df.pivot_table(interest1, context)
+    # accumulates coupling iterations and newton iterations
+    dfe_tsi = df.pivot_table(interest2, context, aggfunc='sum')
+
+    dfe = dfe_ts.merge(dfe_tsi, left_index=True, right_index=True)
+    check_output(dfe, interest, context)
+    return dfe
+
+
+def analysis_simulation(df):
+    interest = ['execution_time']  # 'start_time'
+    context = ['mpi_process']
+    check_input(df, interest, context)
+
+    pt = df.pivot_table(interest, context)
+    check_output(pt, interest, context)
+    return pt
+
+
+def analysis_convergence_newton_iteration(df):
+    dfe_newton_iteration = df.copy()
+    interest = ['dx', 'x', 'dx_x']
+    if 'coupling_iteration' in df:
+        context = ['time_step', 'coupling_iteration', 'process',
+                   'iteration_number', 'component']
+        check_input(df, interest, context)
+        # Eliminate all entries for coupling iteration (not of interest in this study)
+        dfe_newton_iteration['coupling_iteration'] = dfe_newton_iteration.groupby('mpi_process')[
+            ['coupling_iteration']].fillna(method='bfill')
+        dfe_newton_iteration = dfe_newton_iteration[~dfe_newton_iteration['coupling_iteration_process'].notna()]
+        dfe_newton_iteration = dfe_newton_iteration.dropna(subset=['x'])
+
+        pt = dfe_newton_iteration.pivot_table(interest, context)
+
+    else:
+        context = ['time_step', 'process', 'iteration_number', 'component']
+        check_input(df, interest, context)
+        pt = dfe_newton_iteration.pivot_table(interest, context)
+
+    check_output(pt, interest, context)
+    return pt
+
+
+@pre_post_check(interest=['dx', 'x', 'dx_x'], \
+                context=['time_step', 'coupling_iteration', 'coupling_iteration_process',
+                         'component'])
+def analysis_convergence_coupling_iteration(df):
+    # Coupling iteration column will be modified specific for coupling iteration analysis, modified data can not be used for other analysis ->copy!
+    dfe_convergence_coupling_iteration = df.copy()
+    interest = ['dx', 'x', 'dx_x']
+    context = ['time_step', 'coupling_iteration', 'coupling_iteration_process',
+               'component']
+    check_input(df, interest, context)
+
+    dfe_convergence_coupling_iteration['coupling_iteration'] = \
+        dfe_convergence_coupling_iteration.groupby('mpi_process')[['coupling_iteration']].fillna(method='ffill')
+    # All context log lines (iteration_number) have no values for dx, dx_x, x . From now on not needed -> dropped
+    dfe_convergence_coupling_iteration = dfe_convergence_coupling_iteration.dropna(
+        subset=['coupling_iteration_process']).dropna(subset=['x'])
+
+    pt = dfe_convergence_coupling_iteration.pivot_table(interest, context)
+    check_output(pt, interest, context)
+    return pt
+
+
+def time_step_vs_iterations(df):
+    interest = 'iteration_number'
+    context = 'time_step'
+    check_input(df, interest, context)
```

Comment:
This line is causing some trouble. It looks like it is taking every letter as a column.